allennlp/common/tensor.py

```diff
@@ -23,39 +25,39 @@ def get_lengths_from_binary_sequence_mask(mask: torch.ByteTensor):
     return mask.sum(-1).squeeze()
 
 
-def sort_batch_by_length(tensor: torch.FloatTensor, sequence_lengths: torch.LongTensor):
+def sort_batch_by_length(tensor: torch.autograd.Variable, sequence_lengths: torch.autograd.Variable):
     """
     Sort a batch first tensor by some specified lengths.
 
     Parameters
     ----------
-    tensor : torch.FloatTensor, required.
+    tensor : Variable(torch.FloatTensor), required.
         A batch first Pytorch tensor.
-    sequence_lengths : torch.LongTensor, required.
+    sequence_lengths : Variable(torch.LongTensor), required.
         A tensor representing the lengths of some dimension of the tensor which
         we want to sort by.
 
     Returns
     -------
-    sorted_tensor : torch.FloatTensor
+    sorted_tensor : Variable(torch.FloatTensor)
         The original tensor sorted along the batch dimension with respect to sequence_lengths.
-    sorted_sequence_lengths : torch.LongTensor
+    sorted_sequence_lengths : Variable(torch.LongTensor)
         The original sequence_lengths sorted by decreasing size.
-    restoration_indices : torch.LongTensor
-        Indices into the sorted_tensor such that ``sorted_tensor[restoration_indices] == original_tensor``
+    restoration_indices : Variable(torch.LongTensor)
+        Indices into the sorted_tensor such that
+        ``sorted_tensor.index_select(0, restoration_indices) == original_tensor``
     """
-    # Pytorch can't handle sorting with Variables.
-    # TODO(Mark): Assert whether this is allowed - does it break autograd?
-    if isinstance(sequence_lengths, Variable):
-        sequence_lengths = sequence_lengths.data
+
+    if not isinstance(tensor, Variable) or not isinstance(sequence_lengths, Variable):
+        raise ConfigurationError("Both the tensor and sequence lengths must be torch.autograd.Variables.")
 
     sorted_sequence_lengths, permutation_index = sequence_lengths.sort(0, descending=True)
-    sorted_tensor = tensor[permutation_index]
+    sorted_tensor = tensor.index_select(0, permutation_index)
```

Comment:
@DeNeutoy, I was curious if this worked, and so I checked.  Changing this back to `tensor[permutation_index]` still passes tests.  I ran a quick test with `timeit` to see which was faster:

```python
>>> import timeit
>>> timeit.timeit(setup='import torch; a = torch.rand(64, 14, 6); l = torch.rand(64); s, b = l.sort(0, descending=True)', stmt='t = a[b]', number=1000000)
2.71397556500051
>>> timeit.timeit(setup='import torch; a = torch.rand(64, 14, 6); l = torch.rand(64); s, b = l.sort(0, descending=True)', stmt='t = a.index_select(0, b)', number=1000000)
2.9591599720006343
```
Looks like it's about 10% faster to use advanced indexing instead of `index_select`.

Actually, scratch that.  I tried the same thing on the GPU (adding a call to `.cuda()` after the calls to `torch.rand()` in the setup), and the times were basically identical.

So the difference is just readability?  Any other reason to switch, @nelson-liu?

Reply:
Nope, I was just curious (since I would have used advanced indexing). I forgot to follow up, but I believe the reason Mark used `index_select` is because you can't pass a `Variable` directly as advanced indexing input.

Reply:
That may have been the issue back when this was merged, but it works now.  We could switch if we wanted to.  That's good to know for future reference.

Reply:
Sorry I only just saw all this! Yes the variable issue was originally why.